src/SFML/Audio/SoundFileWriterOpus.cpp

```diff
@@ -0,0 +1,259 @@
+
+    // Vendor string
+    memcpy(static_cast<void*>(commentData+12), opusVersion, opusVersionLength);
+
+    // Length of user comments (E.g. you can add a ENCODER tag for SFML)
+    writeUint32(commentData + 12+opusVersionLength, 0);
+
+    op.packet     = commentData;
+    op.bytes      = commentLength;
+    op.b_o_s      = 0;
+    op.e_o_s      = 0;
+    op.granulepos = 0;
+    op.packetno   = m_packageNumber++;
+    ogg_stream_packetin(&m_ogg, &op);
+
+    // This ensures the actual audio data will start on a new page, as per spec
+    flushBlocks();
+
+    return true;
+}
+
+
+////////////////////////////////////////////////////////////
+void SoundFileWriterOpus::write(const Int16* samples, Uint64 count)
+{
+    assert(m_opus);
+
+    const opus_uint32 frame_size = 960;
+    const opus_int32 buffer_size = frame_size * m_channelCount * sizeof(Int16);
+    unsigned char buffer[buffer_size];
+    Uint32 frame_number = 0;
+    Uint8 endOfStream = 0;
+    ogg_packet op;
+
+    while (count > 0)
+    {
+        opus_int32 packet_size;
+
+        // Check if wee need to pad the input
+        if (count < (frame_size * m_channelCount))
+        {
+            Int16 pad[frame_size*m_channelCount];
+            memset(pad, 0, frame_size * m_channelCount);
+            memcpy(pad, samples + (frame_number * frame_size * m_channelCount), count);
+            packet_size = opus_encode(m_opus, pad, frame_size, buffer, buffer_size);
+            endOfStream = 1;
+            count = 0;
+        }
+        else
+        {
+            packet_size = opus_encode(m_opus, samples + (frame_number * frame_size * m_channelCount), frame_size, buffer, buffer_size);
+            count -= frame_size * m_channelCount;
+        }
+
+        if (packet_size < 0)
+        {
+            err() << "An error occurred when encoding sound to opus." << std::endl;
+            break;
+        }
+        op.packet     = buffer;
+        op.bytes      = packet_size;
+        op.granulepos = frame_number * frame_size * 48000ul / m_sampleRate;
+        op.packetno   = m_packageNumber++;
+        op.b_o_s      = 0;
+        op.e_o_s      = endOfStream;
+        ogg_stream_packetin(&m_ogg, &op);
+
+        frame_number++;
+    }
+
+    // Flush any produced block
+    flushBlocks();
+}
+
+
+////////////////////////////////////////////////////////////
+void SoundFileWriterOpus::flushBlocks()
+{
+    ogg_page page;
+    while (ogg_stream_pageout(&m_ogg, &page) > 0 || ogg_stream_flush(&m_ogg, &page) > 0)
+    {
+        m_file.write(reinterpret_cast<const char*>(page.header), page.header_len);
+        m_file.write(reinterpret_cast<const char*>(page.body), page.body_len);
+    }
+}
+
+
+////////////////////////////////////////////////////////////
+void SoundFileWriterOpus::close()
+{
+    if (m_file.is_open())
+    {
+        flushBlocks();
+        m_file.close();
+    }
+
+    // Clear all the ogg/opus structures
+    ogg_stream_clear(&m_ogg);
```

Comment:
Will this crash if `close()` is called twice?

I'm asking because the other variables (`m_file` and `m_opus`) are checked for being deallocated already, but not this one...